---
title: "R Notebook"
output: html_notebook
---

#Setup
```{r}
source("http://bioconductor.org/biocLite.R")
#biocLite()
#biocLite("edgeR")
#biocLite("limma")
#biocLite("Glimma")
library(edgeR)
library(limma)
library(Glimma)
```
#Import data
```{r}
genes_rsem <- read.csv("nematocida.counts.matrix", stringsAsFactors = F, check.names = F)
dim(genes_rsem)
head(genes_rsem)
```

```{r}
counts_matrix <- genes_rsem[,2:13]
rownames(counts_matrix) <- genes_rsem[,1]
dim(counts_matrix)
head(counts_matrix)
```
#Generate DGElist
```{r}
x <- DGEList(counts=counts_matrix, genes = row.names(counts_matrix))
class(x)
x
```
#Organize data
```{r}
samplenames <- substring(colnames(x),7,nchar(colnames(x)))
samplenames
```
```{r}
colnames(x) <- samplenames
treat <- as.factor(c(rep(c("un", "ERTm1", "ERTm2"),3),"ERTm1","un","ERTm2"))
treat <- factor(treat, levels = c("un", "ERTm1", "ERTm2"))
x$samples$group <- treat
lane <- as.factor(rep(c("L004", "L007"), c(9,3)))
x$samples$lane <- lane
x$samples
```
#Data pre-processing
```{r}
cpm <- cpm(x)
lcpm <- cpm(x, log=T)
head(cpm)
```
#Remove lowly expressed genes
First expression checks the number of genes that have zero counts in every condition.
Second expression creates a vector of genes that have cpm >1 in at least four rows. We choose four because that is the number of replicates we have, however, this is an arbitrary number and this cutoff is flexible.
The last expression keeps the genes that match the cutoff described above and refactors the library sizes now that we have dropped a number of genes along with their count data. 
```{r}
table(rowSums(x$counts==0)==12)
keep.exprs <- rowSums(cpm>1)>=4
x <- x[keep.exprs,, keep.lib.sizes=F]
dim(x)
```
#Graph showcasing the filtering
```{r}
library(RColorBrewer)
nsamples <- ncol(x)
col <- brewer.pal(nsamples, 'Paired')
par(mfrow=c(1,2))
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.21), las=2, main="", xlab="")
title(main="Raw data", xlab="log-cpm")
abline(v=0, lty=3)
for (i in 2:nsamples) {
  den <- density(lcpm[,i])
  lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", samplenames, text.col = col, bty="n", cex=0.6)
lcpm <- cpm(x, log=T)
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.21), las=2, main="", xlab="")
title(main="Filtered data", xlab="log-cpm")
abline(v=0, lty=3)
for (i in 2:nsamples) {
  den <- density(lcpm[,i])
  lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", samplenames, text.col = col, bty="n", cex=0.6)
```
#Normalizing gene expression distributions
```{r}
x <- calcNormFactors(x, method = "TMM")
x$samples$norm.factors
```
#Unsupervised clustering
```{r}
lcpm <- cpm(x, log=T)
par(mfrow=c(1,2))
col.group <- treat
levels(col.group) <- brewer.pal(nlevels(col.group), "Set1")
col.group<-as.character(col.group)
col.lane <- lane
levels(col.lane) <- brewer.pal(nlevels(col.lane), "Set2")
col.lane<-as.character(col.lane)
plotMDS(lcpm, labels = treat, col=col.group)
title(main="Treatments")
plotMDS(lcpm, labels = treat, col=col.group, dim=c(3,4))
title(main="Treatments - dim 3,4")
plotMDS(lcpm, labels = lane, col=col.group)
title(main="Lane")
plotMDS(lcpm, labels = lane, dim=c(3,4), col=col.group)
title(main="Lane - dim 3,4")
```
#Differential expression analysis
#Create design matrix and cotnrasts
```{r}
design <- model.matrix(~0+treat+lane)
colnames(design) <- gsub("treat", "", colnames(design))
design
```
```{r}
contrast.matrix <- makeContrasts(
  ERTm1vsN2 = ERTm1 - un,
  ERTm2vsN2 = ERTm2 - un,
  ERTm2vsERTm1 = ERTm2 - ERTm1,
  levels = colnames(design)
  )
contrast.matrix
```
#Remove heteroscedasity
```{r}
par(mfrow=c(1,2))
v <- voom(x, design, plot = T)
vfit <- lmFit(v, design)
vfit <- contrasts.fit(vfit, contrasts = contrast.matrix)
efit <- eBayes(vfit)
plotSA(efit)
```
#Examining differentially expressed genes
I will not be using a logFC criteria in this analysis. If one would like to do so, then use the `treat` function
```{r}
dt <- decideTests(efit)
summary(dt)
```
```{r}
vennDiagram(dt[,1:2], circle.col = c("red", "blue"), include = "up", show.include = T)
```
```{r}
de.common.up <- which(dt[,1] !=0 & dt[,2]!=0)
length(de.common.up)
common.genes <- efit$genes[de.common.up,1]
common.genes
#write.csv(common.genes, "2018_08_03_common_genes_up.csv")
```
```{r}
p <- 0.05
ERTm1.vs.N2 <- topTable(efit, coef = 1, n=Inf, p.value = p)
dim(ERTm1.vs.N2)
head(ERTm1.vs.N2)
ERTm2.vs.N2 <- topTable(efit, coef = 2, n=Inf, p.value = p)
dim(ERTm2.vs.N2)
head(ERTm2.vs.N2)
```
```{r}

```

```{r}
plotMD(efit, column=1, status = dt[,1], main=colnames(efit)[1])
```
```{r}
plotMD(efit, column=2, status = dt[,2], main=colnames(efit)[2])
```
this glMDplot is of ERTm2vsN2
```{r eval=F}
glMDPlot(efit, coef=1, status = dt[,1], main = colnames(efit)[1],
         side.main ="genes", counts = x$counts, groups = treat, launch = F,
         html = "2018_08_03_ERT1vsN2_MDplot")
```

this glMDplot is of ERTm2vsN2
```{r eval=F}
glMDPlot(efit, coef=2, status = dt[,2], main = colnames(efit)[2],
         side.main ="genes", counts = x$counts, groups = treat, launch = F,
         html = "2018_08_03_ERT2vsN2_MDplot")
```
```{r}
sessionInfo()
```

