---
title: "R Notebook"
output:
  html_document:
    df_print: paged
---

#Setup
```{r}
source("http://bioconductor.org/biocLite.R")
#biocLite()
#biocLite("edgeR")
#biocLite("limma")
#biocLite("Glimma")
library(edgeR)
library(limma)
library(Glimma)
```
#Import data
```{r}
genes_rsem <- read.csv("nematocida.counts.matrix", stringsAsFactors = F, check.names = F)
dim(genes_rsem)
head(genes_rsem)
```

```{r}
counts_matrix <- genes_rsem[,2:10]
rownames(counts_matrix) <- genes_rsem[,1]
colnames(counts_matrix)
dim(counts_matrix)
head(counts_matrix)
```
#Generate DGElist
```{r}
x <- DGEList(counts=counts_matrix, genes = row.names(counts_matrix))
class(x)
x
```
#Organize data
```{r}
samplenames <- substring(colnames(x),7,nchar(colnames(x)))
samplenames
```
```{r}
colnames(x) <- samplenames
treat <- as.factor(rep(c("un", "ERTm1", "ERTm2"),3))
treat <- factor(treat, levels = c("un", "ERTm1", "ERTm2"))
x$samples$group <- treat
x$samples
```
#Data pre-processing
```{r}
cpm <- cpm(x)
lcpm <- cpm(x, log=T)
head(cpm)
```
#Remove lowly expressed genes
First expression checks the number of genes that have zero counts in every condition.
Second expression creates a vector of genes that have cpm >1 in at least three rows. We choose three because that is the number of replicates we have, however, this is an arbitrary number and the cutoff is flexible.
The last expression keeps the genes that match the cutoff described above and refactors the library sizes now that we have dropped a number of genes along with their count data. 
```{r}
table(rowSums(x$counts==0)==12)
keep.exprs <- rowSums(cpm>1)>=3
x <- x[keep.exprs,, keep.lib.sizes=F]
dim(x)
```
#Graph showcasing the filtering
```{r}
library(RColorBrewer)
nsamples <- ncol(x)
col <- brewer.pal(nsamples, 'Paired')
par(mfrow=c(1,2))
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.21), las=2, main="", xlab="")
title(main="Raw data", xlab="log-cpm")
abline(v=0, lty=3)
for (i in 2:nsamples) {
  den <- density(lcpm[,i])
  lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", samplenames, text.col = col, bty="n", cex=0.6)
lcpm <- cpm(x, log=T)
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.21), las=2, main="", xlab="")
title(main="Filtered data", xlab="log-cpm")
abline(v=0, lty=3)
for (i in 2:nsamples) {
  den <- density(lcpm[,i])
  lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", samplenames, text.col = col, bty="n", cex=0.6)
```
#Normalizing gene expression distributions
```{r}
x <- calcNormFactors(x, method = "TMM")
x$samples$norm.factors
```
#Unsupervised clustering
```{r}
lcpm <- cpm(x, log=T)
par(mfrow=c(1,2))
col.group <- treat
levels(col.group) <- brewer.pal(nlevels(col.group), "Set1")
col.group<-as.character(col.group)
plotMDS(lcpm, labels = treat, col=col.group)
title(main="Treatments")
plotMDS(lcpm, labels = treat, col=col.group, dim=c(3,4))
title(main="Treatments - dim 3,4")
```

#Differential expression analysis
Create design matrix and cotnrasts
```{r}
design <- model.matrix(~0+treat)
colnames(design) <- gsub("treat", "", colnames(design))
design
```
```{r}
contrast.matrix <- makeContrasts(
  ERTm1vsN2 = ERTm1 - un,
  ERTm2vsN2 = ERTm2 - un,
  levels = colnames(design)
  )
contrast.matrix
```
#Remove heteroscedasity
```{r}
par(mfrow=c(1,2))
v <- voom(x, design, plot = T)
vfit <- lmFit(v, design)
vfit <- contrasts.fit(vfit, contrasts = contrast.matrix)
efit <- eBayes(vfit)
plotSA(efit)
```
#Examining differentially expressed genes
I will not be using a logFC criteria in this analysis. If one would like to do so, then use the `treat` function
```{r}
dt <- decideTests(efit)
summary(dt)
```
```{r}
vennDiagram(dt[,1:2], circle.col = c("red", "blue"), include = "up", show.include = T)
```
```{r}
de.common.up <- which(dt[,1] !=0 & dt[,2]!=0)
length(de.common.up)
common.genes <- efit$genes[de.common.up,1]
common.genes
#write.csv(common.genes, "2018_08_03_common_genes_up.csv")
```
```{r}
p <- 0.05
ERTm1.vs.N2 <- topTable(efit, coef = 1, n=Inf, p.value = p)
dim(ERTm1.vs.N2)
head(ERTm1.vs.N2)
ERTm2.vs.N2 <- topTable(efit, coef = 2, n=Inf, p.value = p)
dim(ERTm2.vs.N2)
head(ERTm2.vs.N2)
```
Clustering heatmap for ERTm1 genes (all 87 UP genes)
```{r}
#fix margins too large error
par(mar=c(2,2,2,2))
par(mfrow=c(1,1))
#Heatmap
library(gplots)
i <- which(v$genes$genes %in% ERTm1.vs.N2[,1])
mycol <- colorpanel(1000,"blue","white","red")
heatmap.2(v$E[i,], scale = "row", labRow = v$genes$genes[i], labCol = treat,
          col = mycol, trace = "none", density.info = "none", margins = c(4,3), keysize = 1,
          lhei = c(2,4), dendrogram = "column", cexRow = 0.55, cexCol = 0.67)
```

```{r}
plotMD(efit, column=1, status = dt[,1], main=colnames(efit)[1])
```
```{r}
plotMD(efit, column=2, status = dt[,2], main=colnames(efit)[2])
```
this glMDplot is of ERTm2vsN2
```{r eval=F}
glMDPlot(efit, coef=1, status = dt[,1], main = colnames(efit)[1],
         side.main ="genes", counts = x$counts, groups = treat, launch = F,
         html = "2018_08_06_ERT1vsN2_MDplot")
```

this glMDplot is of ERTm2vsN2
```{r eval=F}
glMDPlot(efit, coef=2, status = dt[,2], main = colnames(efit)[2],
         side.main ="genes", counts = x$counts, groups = treat, launch = F,
         html = "2018_08_06_ERT2vsN2_MDplot")
```

#Extracting gene lists
```{r}
#Create vector containing commone UP gene names
de.common.up <- which(dt[,1] !=0 & dt[,2]!=0)
length(de.common.up)
common.genes <- efit$genes[de.common.up,1]
common.genes
#collect names and logFC values of common UP genes
ERTm1.vs.N2.common <- ERTm1.vs.N2[common.genes,2]
ERTm2.vs.N2.common <- ERTm2.vs.N2[common.genes,2]
de <- data.frame(ERTm1.vs.N2.common, ERTm2.vs.N2.common, row.names = ERTm1.vs.N2[common.genes,1])
de
```
```{r}
#Collect unique genes from both sets
#Unique for ERTm1
de.ertm1.unique <- which(dt[,1] !=0 & dt[,2]==0)
length(de.ertm1.unique)
ERTm1.vs.N2.unique <- efit$genes[de.ertm1.unique,1]
ERTm1.vs.N2.unique <- ERTm1.vs.N2[ERTm1.vs.N2.unique,1:2]
ERTm1.vs.N2.unique
#uniqe for ERTm2
de.ertm2.unique <- which(dt[,1] ==0 & dt[,2]!=0)
length(de.ertm2.unique)
ERTm2.vs.N2.unique <- efit$genes[de.ertm2.unique,1]
ERTm2.vs.N2.unique <- ERTm2.vs.N2[ERTm2.vs.N2.unique,1:2]
ERTm2.vs.N2.unique
```
#Write to xlsx
```{r}
library(openxlsx)

wb <- createWorkbook()
addWorksheet(wb, sheetName = "Common UP")
addWorksheet(wb, sheetName = "ERTm1 unique")
addWorksheet(wb, sheetName = "ERTm2 unique")

writeData(wb, "Common UP", de, rowNames = T)
setColWidths(wb, sheet = 1, cols = 1:3, widths = "auto")
writeData(wb, "ERTm1 unique", ERTm1.vs.N2.unique)
setColWidths(wb, sheet = 2, cols = 1:2, widths = "auto")
writeData(wb, "ERTm2 unique", ERTm2.vs.N2.unique)
setColWidths(wb, sheet = 3, cols = 1:2, widths = "auto")

saveWorkbook(wb, "2018_08_06_Nematocida_DEGenes.xlsx", overwrite = T)
```


```{r}
sessionInfo()
```

